graph/layout/eades.go

```diff
@@ -0,0 +1,171 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package layout
+
+import (
+	"math"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/spatial/barneshut"
+	"gonum.org/v1/gonum/spatial/r2"
+)
+
+// EadesR2 implements the graph layout algorithm essentially as
+// described in "A heuristic for graph drawing", Congressus
+// numerantium 42:149-160.
+// The implementation here uses the Barnes-Hut approximation for
+// global repulsion calculation, and edge weights are considered
+// when calculating adjacent node attraction.
+type EadesR2 struct {
+	// Updates is the number of updates to perform.
+	Updates int
```

Comment:
For me this does not fit here, the caller will be calling Update in a loop anyway, so they can control how many updates are performed. Optimizer looks like a better place? Or are you including it here because you don't want to assume all layout algorithms being iterative? But in that case why pass them through Optimizer? I don't want to overthink this and replicate what `optimize` does (although on the other hand, it almost looks like this could be adapted to use `optimize` :-) ).

Reply:
This is indeed here because there are layout algorithms that do not do iterative updates. A layout that I want to include is minumum edge crossing circular layout (e.g. [here](https://www.sciencedirect.com/science/article/pii/S1570866705000031/pdfft?md5=9012fb96c356e2a2611e6f9d32bb6ece&pid=1-s2.0-S1570866705000031-main.pdf). The reason to pass them through optimiser is to allow a general API for constructing layouts. The user does not need to know the type approach taken.